[HLSL] Add `Increment`/`DecrementCounter` methods to structured buffers

hekota opened the pull request:
Introduces `__builtin_hlsl_buffer_update_counter` clang buildin that is used to implement the `IncrementCounter` and `DecrementCounter` methods on `RWStructuredBuffer` and `RasterizerOrderedStructuredBuffer` (see Note).

The builtin is translated to LLVM intrisic `llvm.dx.bufferUpdateCounter` or `llvm.spv.bufferUpdateCounter`.

Introduces `BuiltinTypeMethodBuilder` helper in `HLSLExternalSemaSource` that enables adding methods to builtin types using builder pattern like this:
```
   BuiltinTypeMethodBuilder(Sema, RecordBuilder, "MethodName", ReturnType)
       .addParam("param_name", Type, InOutModifier)
       .callBuiltin("buildin_name", { BuiltinParams })
       .finalizeMethod();
```

Fixes #113513

hekota committed:
[HLSL] Add Increment/DecrementCounter methods to structured buffers

Introduces `__builtin_hlsl_buffer_update_counter` clang buildin that is used to implement IncrementCounter and DecrementCounter methods on RWStructuredBuffer and RasterizerOrderedStructuredBuffer. The builtin is translated to LLVM intrisics llvm.dx.bufferUpdateCounter/llvm.spv.bufferUpdateCounter.

Introduces `BuiltinTypeMethodBuilder` helper in `HLSLExternalSemaSource` that allows adding methods to builtin types
using the builder pattern like this:

   BuiltinTypeMethodBuilder(Sema, RecordBuilder, "MethodName", ReturnType)
       .addParam("param_name", Type, InOutModifier)
       .callBuiltin("buildin_name", { BuiltinParams })
       .finalizeMethod();

Fixes #113513

llvm-beanz commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -528,8 +719,13 @@ void HLSLExternalSemaSource::defineHLSLTypesWithForwardDeclarations() {
                     ResourceKind::TypedBuffer, /*IsROV=*/false,
                     /*RawBuffer=*/true)
         .addArraySubscriptOperators()
+        .addIncrementCounterMethod(*SemaPtr)
```
Should we make the BuiltinTypeDeclBuilder store a reference to Sema to avoid needing to pass these pointers around?

llvm-beanz commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -359,6 +356,176 @@ struct TemplateParameterListBuilder {
+    if (DeclBuilder.Record->isCompleteDefinition())
+      return DeclBuilder;
+
+    if (!Method)
```
This and the check above both seem like they would be errors. Maybe they should be asserts?

hekota replied:
I copied this approach from the way `BuiltinTypeDeclBuilder` currently does it. I will change these to asserts everywhere.

hekota replied:
The other check (`!Method`) is not quite an error, it just means no body building method have been called (the created method has no body).

hekota replied:
Actually, the check `if (DeclBuilder.Record->isCompleteDefinition())` needs to stay in. We might already have a complete definition from a precompiled header.

If I replace these with asserts the precompiled header test `C:\llvm-project3\clang\test\AST\HLSL\pch_with_buf.hlsl` fails.

llvm-beanz replied:
If the PCH is providing a completed definition, we probably shouldn't be calling into the complete callback since all of that stuff should fail. Can we detect that earlier and never get here?

llvm-beanz replied:
> The other check (`!Method`) is not quite an error, it just means no body building method have been called (the created method has no body).

A method with no body seems like a logic error on the person writing code (hence the request for an assert). Why would we want to enable generating empty methods?

llvm-beanz commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -359,6 +356,176 @@ struct TemplateParameterListBuilder {
+    if (!Method->hasBody()) {
+      ASTContext &AST = S.getASTContext();
+      if (ReturnTy != AST.VoidTy && !StmtsList.empty()) {
+        if (Expr *LastExpr = dyn_cast<Expr>(StmtsList.back())) {
```
Do we need to handle the case where `LastExpr->getType() != ReturnTy`?
If not, maybe we should add an assert something like:
```
assert(AST.hasSameUnqualifiedType(LastExpr->getType(), ReturnTy ) && "Should be the same type!");
```

llvm-beanz commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -282,6 +258,23 @@ struct BuiltinTypeDeclBuilder {
+  }
+
+  QualType getFirstTemplateTypeParam() {
+    if (Template) {
```
This should be an assert rather than an `if`, this method should never get called if Template isn't set.

llvm-beanz commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -282,6 +258,23 @@ struct BuiltinTypeDeclBuilder {
 
+  FieldDecl *getResourceHandleField() {
+    FieldDecl *FD = Fields["h"];
+    if (FD && FD->getType()->isHLSLAttributedResourceType())
```
These should be `asserts` rather than `if`.

llvm-beanz reviewed (commented):
Mostly looking good. I really like the additions to the builder API. Can you please add an AST test to verify the shape of the new AST nodes and their instantiations?

hekota committed:
Code review feedback - add test & asserts, add Sema to main decl builder, fix generated CallExpr type

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into sb4-counter

hekota committed:
Code review feedback - assert of completed definition or no method body

Do not complete definition if we already got one from precompiled headers.

hekota committed:
clang-format

llvm-beanz commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -354,27 +345,230 @@ struct TemplateParameterListBuilder {
+  BuiltinTypeMethodBuilder &
+  callBuiltin(StringRef BuiltinName, ArrayRef<Expr *> CallParms,
+              bool AddResourceHandleAsFirstArg = true) {
+    if (!Method)
```
nit: Might be nice to have a comment here explaining the need for the create call.
```suggestion
    // The first statement added to a method creates the declaration.
    if (!Method)
```

llvm-beanz commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -354,27 +345,230 @@ struct TemplateParameterListBuilder {
+    // FIXME: Call the buildin with all of the method parameters
+    // plus optional resource handle as the first arg.
+    llvm_unreachable("not yet implemented");
+  }
```
Since this is unused, maybe we should remove it from this PR and add it in the subsequent PR when it is used.

llvm-beanz reviewed (approved):
A couple small comment take or leave, but looks good.

hekota commented on clang/include/clang/Basic/Builtins.td:
```diff
@@ -4846,7 +4846,6 @@ def HLSLSaturate : LangBuiltin<"HLSL_LANG"> {
   let Prototype = "void(...)";
 }
 
-
```
This is intentional - I'm removing an extra empty line.

hekota committed:
Code review feedback - remove unused method, add comment

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into sb4-counter

hekota committed:
Add Increment/DecrementCounter to RasterizerOrderedStructuredBuffer; update tests

pow2clk commented on HLSLExternalSemaSource.cpp:
```diff
@@ -0,0 +1,771 @@
+//===--- HLSLExternalSemaSource.cpp - HLSL Sema Source --------------------===//
```
I don't think this file should be here.

hekota replied:
Oops! :)

pow2clk commented on clang/include/clang/Basic/Builtins.td:
```diff
@@ -4883,6 +4882,12 @@ def HLSLRadians : LangBuiltin<"HLSL_LANG"> {
 
+def HLSLBufferUpdateCounter : LangBuiltin<"HLSL_LANG"> {
+  let Spellings = ["__builtin_hlsl_buffer_update_counter"];
+  let Attributes = [NoThrow, Const];
```
I don't know what effect these attributes have or if I'm understanding Const correctly, but if that implies it doesn't change any memory, it does.

pow2clk reviewed (commented):
#113648 is merged. Should probably update the description as I think the corresponding inclusion of ROV buffers went in too.

I haven't finished my review, but I thought I'd get these two points in quickly.

pow2clk commented on llvm/include/llvm/IR/IntrinsicsDirectX.td:
```diff
@@ -35,6 +35,9 @@ def int_dx_typedBufferLoad_checkbit
     : DefaultAttrsIntrinsic<[], [llvm_any_ty, llvm_i32_ty, llvm_anyvector_ty]>;
 
+def int_dx_bufferUpdateCounter
+    : DefaultAttrsIntrinsic<[llvm_i32_ty], [llvm_any_ty, llvm_i32_ty], [IntrWriteMem]>;
```
`IntrWriteMem` indicates that this only writes memory. Since it returns the original counter value, it does read some memory too.

tex3d replied:
No attribute would indicate may read/write memory.  However, what about `IntrInaccessibleMemOrArgMemOnly`?  I wonder if that attribute would more accurately capture what we want.  This counter memory is unique and cannot alias any other memory operations in the module.

pow2clk commented on clang/test/SemaHLSL/BuiltIns/buffer_update_counter-errors.hlsl:
```diff
@@ -0,0 +1,22 @@
+  __builtin_hlsl_buffer_update_counter(res, x);
+
+  // expected-error@+1 {{argument 1 must be constant integer 1 or -1}}
+  __builtin_hlsl_buffer_update_counter(res, 10);
```
Might want a test that integer values are required since there is a unique message and check function for it. Might also want a test that proper usage of `__builtin_hlsl_buffer_update_counter` produces no diagnostics.

pow2clk commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -271,53 +246,70 @@ struct BuiltinTypeDeclBuilder {
-    if (Builder.Record->isCompleteDefinition())
-      return *this;
+    assert(!Builder.Record->isCompleteDefinition() &&
+           "record is already complete");
```
I approve of altering the expectations the callers of this function should have. I think the previous behavior was less a convenience than an unpleasant surprise.

pow2clk commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -271,53 +246,70 @@ struct BuiltinTypeDeclBuilder {
-    if (Record->isCompleteDefinition())
-      return *this;
+    // we might already have complete definition from a precompiled header
+    assert(!Record->isCompleteDefinition() && "record is already complete");
```
If we might already have the definition, then we shouldn't assert here.

hekota committed:
Remove accidentally added file; update builtin to not be Const

pow2clk commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -343,27 +336,224 @@ struct TemplateParameterListBuilder {
+// method that builds the body is called it creates the CXXMethodDecl and
+// ParmVarDecls instances. These can then be referenced from the body building
+// methods. Destructor or an explicit call to finalizeMethod() will complete
+// the method definition.
```
Great description!

tex3d commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -343,27 +336,224 @@ struct TemplateParameterListBuilder {
+// using the builder pattern like this:
+//
+//   BuiltinTypeMethodBuilder(Sema, RecordBuilder, "MethodName", ReturnType)
+//       .addParam("param_name", Type, InOutModifier)
```
I don't see anything adding or using parameters here.  I also don't know how you'd use a parameter you added this way in the builtin call params you pass in (how do you reference it to create the DeclRefExpr?).

Since this part is unused, should we really be adding it at this time?  Could we add it when it's used to make sure it's working as intended?

hekota replied:
I'll keep the `addParam` method decl since it is mentioned here in the comment but replace the body with `llvm_unreachable`. We can add the implementation when it can be tested.

tex3d commented on clang/include/clang/Basic/Builtins.td:
```diff
@@ -4883,6 +4882,12 @@ def HLSLRadians : LangBuiltin<"HLSL_LANG"> {
+def HLSLBufferUpdateCounter : LangBuiltin<"HLSL_LANG"> {
+  let Spellings = ["__builtin_hlsl_buffer_update_counter"];
+  let Attributes = [NoThrow];
+  let Prototype = "uint32_t(...)";
```
Why does this look like it's overloaded?  Shouldn't this have a specific signature like `"uint32_t(__hlsl_resource_t, int)"`?

hekota replied:
It is overloaded. The intrinsic takes `__hlsl_resource_t` decorated with a resource type attributes and every unique combination of the attributes means a different resource type. And I don't think the attributes can be expressed here anyway even if there was just one overload.

tex3d commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -109,22 +116,16 @@ struct BuiltinTypeDeclBuilder {
     ASTContext &Ctx = S.getASTContext();
     TypeSourceInfo *ElementTypeInfo = nullptr;
 
     QualType ElemTy = Ctx.Char8Ty;
```
Pre-existing side-note: Does this mean that `ByteAddressBuffer` will have an element type of `i8`?  Shouldn't it be `void`, to clearly differentiate this case?

hekota replied:
`i8` seems like a good default to me. We can decide if we want to use `void` instead when we implement the `ByteAddressBuffer`.

tex3d commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -271,53 +246,70 @@ struct BuiltinTypeDeclBuilder {
   }
 
+  FieldDecl *getResourceHandleField() {
+    FieldDecl *FD = Fields["h"];
```
Won't this add "h" when accessed this way?
How about:
```suggestion
    FieldDecl *FD = Fields.lookup("h");
```

hekota replied:
Both [] operator and lookup will add the field if it is not already there. I will update it to use "find".

tex3d replied:
I thought `lookup` was right:
https://github.com/llvm/llvm-project/blob/30af6fb163add17a6be515200881afdff91d213a/llvm/include/llvm/ADT/StringMap.h#L251-L258

hekota replied:
Yeah, it would probably work fine too, if default value `ValueTy()` just means `nullptr` when `ValueTy` is a pointer.
I looked at the `lookup` documentation and it says: "Return the entry for the specified key, or a default constructed value if no such entry exists."

tex3d commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -343,27 +336,224 @@ struct TemplateParameterListBuilder {
+// method that builds the body is called it creates the CXXMethodDecl and
+// ParmVarDecls instances. These can then be referenced from the body building
+// methods. Destructor or an explicit call to finalizeMethod() will complete
+// the method definition.
```
There are some built-in behaviors/assumptions, some of which probably should be mentioned in the documentation comment.

Here are some oddities I noticed:
- The value returned from the last (statement) expression (builtin call only for now) will be used as the return value for the method (type from builtin must match `ReturnType`).  I get the convenience for this use case, and maybe we never need anything more, but it does seem like we should point this out in the comment.
- It looks like it's trying to be flexible to allow multiple calls to be added, but I don't know any way this architecture would allow you to capture returned values from those calls, reference those in subsequent calls, reference method args as mentioned earlier, etc...  It seems like the interface has been designed tightly around the one use case, but has hints of incompletely designed flexibility (which would likely require changes to this design to take advantage of).
- `callBuiltin()`: Default passing of MemberExpr for the resource handle of `this` from RecordBuilder as the first argument.  I feel like this could be made more explicit, such as by adding a method `BuiltinTypeDeclBuilder::getResourceHandleExpr()`, and a modification of BuiltinParams specified when used below to: `{getResourceHandleExpr(), One}`.

hekota replied:
I am trying to keep the same builder pattern for the interface as we already have in the `BuiltinTypeDeclBuilder`. I did have more than just one use case in mind I designed it, but at this moment I can only add building methods that can be tested as part of task llvm/llvm-project#113513. We are going to add more building methods as needed as we implement the rest of the HLSL methods.

For example this is how the `Append(T value)` method on the `AppendStructuredBuffer` would look like:

```
BuiltinTypeDeclBuilder &BuiltinTypeDeclBuilder::addIncrementCounterMethod() {
  ASTContext &AST = S.getASTContext();
  Expr *One = IntegerLiteral::Create(...);
  return BuiltinTypeMethodBuilder(S, *this, "Append", AST.VoidTy)
      .addParam("value", getFirstTemplateTypeParam())
      .callBuiltin("__builtin_hlsl_buffer_update_counter", {One})
      .callBuiltinForwardParams("__builtin_hlsl_buffer_load") 
      .finalizeMethod();
}
```
The `callBuiltinForwardParams` passes in the resource handle as the first argument of the builtin (unless that is not desired and the method takes a bool to disable that) and then it adds/forwards all of the Append method params to the builtin call (in this the case the `value`). I believe `callBuiltinForwardParams` will probably be the most commonly used case when we implement other HLSL methods.

Without the implicit behavior it might look like this:
```
BuiltinTypeDeclBuilder &BuiltinTypeDeclBuilder::addAppendMethod() {
  ASTContext &AST = S.getASTContext();
  Expr *One = IntegerLiteral::Create(...);
  BuiltinTypeMethodBuilder MB =
      BuiltinTypeMethodBuilder(S, *this, "Append", AST.UnsignedIntTy);
  return MB
      .addParam("value", getFirstTemplateTypeParam())
      .addStmt(callBuiltin(S, "__builtin_hlsl_buffer_update_counter",
                           {MB.getResourceHandleExpr(), One}))
      .returnExpr(callBuiltin(S, "__builtin_hlsl_buffer_load",
                              {MB.getResourceHandleExpr(), MB.getMethodArg(0)}))
}
```
I think the first case looks much simpler. As long as the implicit behavior is well documented and used often enough, I would prefer to keep it like that. I will add more comments pointing out the implicit behavior.

hekota replied:
@tex3d - Justin is looking into having `callBuiltin` take placeholder values (something like `std::placeholders`) that would be used to reference the handle and method arguments.

@bogner FYI

tex3d commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -343,27 +336,224 @@ struct TemplateParameterListBuilder {
+
+    if (!Method->hasBody()) {
+      ASTContext &AST = DeclBuilder.S.getASTContext();
+      if (ReturnTy != AST.VoidTy && !StmtsList.empty()) {
```
Shouldn't we assert for this case?
```suggestion
      assert(ReturnTy == AST.VoidTy || !StmtsList.empty() &&
             "nothing to return from non-void method");
      if (ReturnTy != AST.VoidTy && !StmtsList.empty()) {
```

tex3d commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -343,27 +336,224 @@ struct TemplateParameterListBuilder {
+      SourceLocation());
+  return BuiltinTypeMethodBuilder(S, *this, "IncrementCounter",
+                                  AST.UnsignedIntTy)
+      .callBuiltin("__builtin_hlsl_buffer_update_counter", {One})
```
I think if we were to generalize this a bit more, we'd end up with something like:
`.returnExpr(CallBuiltin(S, "__builtin_hlsl_buffer_update_counter", {getResourceHandleExpr() One})`

Then we wouldn't have to depend on extra implicit functionality in 'callBuiltin' and `finalizeMethod`.

hekota replied:
See my comments above.

tex3d commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -553,3 +748,20 @@ void HLSLExternalSemaSource::CompleteType(TagDecl *Tag) {
+  // AllowBuiltinCreation is false but LookupDirect will create
+  // the builtin when searching the global scope anyways...
+  S.LookupName(R, S.getCurScope());
+  // FIXME: If the builtin function was user-declared in global scope,
```
Wouldn't that require doing so here inside HLSLExternalSemaSource?  How else would this find something before we even start parsing the built-in HLSL header, let alone the source file?

tex3d replied:
Since this is just moving a pre-existing function, feel free to ignore this comment for the purpose of completing the PR.  I'm just wondering out loud.

tex3d commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -2161,6 +2186,22 @@ bool SemaHLSL::CheckBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
     break;
   }
+  case Builtin::BI__builtin_hlsl_buffer_update_counter: {
+    if (SemaRef.checkArgCount(TheCall, 2) ||
+        CheckResourceHandle(&SemaRef, TheCall, 0) ||
+        CheckInt(&SemaRef, TheCall, 1))
+      return true;
```
Shouldn't this be a specific int type (not an integer type of any size/signedness)?  If so, couldn't we have a check function that takes a single type and call it like `CheckArgType(&SemaRef, TheCall, 1, SemaRef.Context.IntTy)`?

Or perhaps it could just be checked and diagnosed here instead?
```suggestion
    if (SemaRef.checkArgCount(TheCall, 2) ||
        CheckResourceHandle(&SemaRef, TheCall, 0))
      return true;
    if (TheCall->getArg(ArgIndex)->getType() != SemaRef.Context.IntTy) {
      S->Diag(TheCall->getArg(0)->getExprLoc(), diag::err_typecheck_expect_int)
        << ArgType;
      return true;
    }
```

tex3d commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -1880,6 +1880,31 @@ static bool CheckVectorSelect(Sema *S, CallExpr *TheCall) {
+  assert(TheCall->getNumArgs() >= ArgIndex);
+  QualType ArgType = TheCall->getArg(ArgIndex)->getType();
+  if (!ArgType->isIntegerType()) {
+    S->Diag(TheCall->getArg(0)->getBeginLoc(), diag::err_typecheck_expect_int)
```
What about `getExprLoc` instead of `getBeginLoc`
```suggestion
    S->Diag(TheCall->getArg(0)->getExprLoc(), diag::err_typecheck_expect_int)
```

pow2clk replied:
At any rate, I think the `0` should be `ArgIndex`

tex3d commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -2161,6 +2186,22 @@ bool SemaHLSL::CheckBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
     break;
   }
+  case Builtin::BI__builtin_hlsl_buffer_update_counter: {
+    if (SemaRef.checkArgCount(TheCall, 2) ||
+        CheckResourceHandle(&SemaRef, TheCall, 0) ||
+        CheckInt(&SemaRef, TheCall, 1))
+      return true;
+    Expr *OffsetExpr = TheCall->getArg(1);
+    std::optional<llvm::APSInt> Offset =
+        OffsetExpr->getIntegerConstantExpr(SemaRef.getASTContext());
+    if (!Offset.has_value() || abs(Offset->getExtValue()) != 1) {
+      SemaRef.Diag(TheCall->getArg(1)->getBeginLoc(),
```
```suggestion
      SemaRef.Diag(TheCall->getArg(1)->getExprLoc(),
```

tex3d replied:
Shouldn't this be a specific int type (not an integer type of any size/signedness)?  If so, couldn't we have a check function that takes a single type and call it like `CheckArgType(&SemaRef, TheCall, 1, SemaRef.Context.IntTy)`?

Or perhaps it could just be checked and diagnosed here instead?
```suggestion
    if (SemaRef.checkArgCount(TheCall, 2) ||
        CheckResourceHandle(&SemaRef, TheCall, 0))
      return true;
    Expr *OffsetExpr = TheCall->getArg(1);
    SourceLocation OffsetLoc = OffsetExpr->getExprLoc();
    if (OffsetExpr->getType() != SemaRef.Context.IntTy) {
      S->Diag(OffsetLoc, diag::err_typecheck_expect_int) << ArgType;
      return true;
    }
    std::optional<llvm::APSInt> Offset =
        OffsetExpr->getIntegerConstantExpr(SemaRef.getASTContext());
    if (!Offset.has_value() || abs(Offset->getExtValue()) != 1) {
      SemaRef.Diag(OffsetLoc,
```

tex3d commented on clang/test/SemaHLSL/BuiltIns/buffer_update_counter-errors.hlsl:
```diff
@@ -0,0 +1,22 @@
+// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.6-library %s -emit-llvm-only -disable-llvm-passes -verify
+
+using handle_t = __hlsl_resource_t [[hlsl::resource_class(UAV)]] [[hlsl::contained_type(int)]];
```
Should we be ensuring the resource attributes are compatible with the operation?  This should be a `RWStructuredBuffer` only: UAV, raw and with contained_type.  Like diagnostics for each of the `bad_handle_*` cases below:
```suggestion
// RWStructuredBuffer<int>
using handle_t = __hlsl_resource_t [[hlsl::resource_class(UAV)]] [[hlsl::contained_type(int)]] [[hlsl::raw_buffer]];
// RWBuffer<int>
using bad_handle_not_raw_t = __hlsl_resource_t [[hlsl::resource_class(UAV)]] [[hlsl::contained_type(int)]];
// RWByteAddressBuffer
using bad_handle_no_type_t = __hlsl_resource_t [[hlsl::resource_class(UAV)]] [[hlsl::raw_buffer]];
// StructuredBuffer
using bad_handle_not_uav_t = __hlsl_resource_t [[hlsl::resource_class(SRV)]] [[hlsl::contained_type(int)]] [[hlsl::raw_buffer]];
```

hekota committed:
Add comments, make getResourceHandleExpr public, add const int helper to simplify method building code

hekota committed:
code review feedback

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into sb4-counter

hekota committed:
fix merge

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into sb4-counter

hekota committed:
Update intrinsics name and arguments to match DXIL and SPIRV.

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into sb4-counter

hekota committed:
Add bitcast from builtin i32 param to intrinsic's i8

Few changes to minimize conflicts with another PR in progress.
Update handle field name and tests.

bogner reviewed (approved):
Just a few nitpicks - this looks great, thanks!

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into sb4-counter

bogner commented on clang/lib/Sema/HLSLExternalSemaSource.cpp:
```diff
@@ -12,44 +12,51 @@
 struct TemplateParameterListBuilder;
 
 struct BuiltinTypeDeclBuilder {
+  Sema &S;
```
Can we use a longer name for this? Single variable member names are uncomfortably likely to collide with local names and cause confusion.

hekota committed:
code review feedback, fix merge, remove unused addConceptSpecializationExpr declaration

hekota committed:
Rename S to SemaRef & clang-format

hekota committed:
Merge branch 'main' of https://github.com/llvm/llvm-project into sb4-counter
